### PR TITLE
Generate clients, fakeclients, informers and listers for go consumers of cartographer to import

### PR DESCRIPTION
**Pull request by rawlingsj**
Initial stab and generating clients etc that kube builder doesn't generate.  This is just an initial attempt which we can discuss / tidy up/ work on.

/cc @kauana @cirocosta @jsanin

### CODE REVIEW

File: pkg/apis/carto/v1alpha1/cluster_config_template_test.go [hunk 1]
```diff
@@ -17,12 +17,11 @@ package v1alpha1_test
 import (
 	"encoding/json"
 
+	"github.com/vmware-tanzu/cartographer/pkg/apis/carto/v1alpha1"
```
Comment (emmjohnson) [inline pkg/apis/carto/v1alpha1/cluster_config_template_test.go]: the is out of order

File: pkg/apis/carto/v1alpha1/cluster_delivery_test.go [hunk 2]
```diff
@@ -17,37 +17,36 @@ package v1alpha1_test
 import (
 	"reflect"
 
+	v1alpha12 "github.com/vmware-tanzu/cartographer/pkg/apis/carto/v1alpha1"
```
Comment (cirocosta) [inline pkg/apis/carto/v1alpha1/cluster_delivery_test.go]: oh weird, apparently a bunch of places got this replacement 🤔 is that coming from the generators included here? an IDE? iirc goland tries to be smart about these things and sometimes gets it wrong

File: pkg/apis/carto/v1alpha1/cluster_config_template_test.go (hunk 1 above)
Comment (cirocosta) [reply pkg/apis/carto/v1alpha1/cluster_config_template_test.go]: yeah, interesting how it is preserving the notion of local imports but it's enforcing having it in this specific location

**Comment by emmjohnson** [comment]
Hey @rawlingsj it looks like the imports got messed up with the changes. It would be nice to revert those changes so we can see the actual changes.

**Comment by cirocosta** [comment]
also, is it common practice to include the fakes? cartographer itself doesn't have any need for them as is - would it be enough to expose the listers/informers/etc but not these fakes for the custom resources?

**Comment by kauana** [comment]
@cirocosta Not sure if this is common practice, but this is what Knative [does](https://github.com/knative/serving/tree/5a51323d83e2b9dfccd8228e4f8604136831a902/pkg/client/clientset/versioned/fake).

**Comment by evankanderson** [comment]
> also, is it common practice to include the fakes? cartographer itself doesn't have any need for them as is - would it be enough to expose the listers/informers/etc but not these fakes for the custom resources?

I think the fakes make it easier for clients to test, rather than for Cartographer itself.

Many of these typed clients are particularly intended for use by other components looking to consume Cartographer. Hopefully, go 1.18's type parameters ("generics") will reduce the need for this sort of boilerplate code generation.

File: pkg/apis/carto/v1alpha1/cluster_delivery_test.go (hunk 2 above)
Comment (rawlingsj) [reply pkg/apis/carto/v1alpha1/cluster_delivery_test.go]: yeah it's a little annoying, as you say it was a folder move in goland which auto updated the package imports, will push a better version shortly

**Comment by rawlingsj** [comment]
Just force pushed to my branch so that the changes are clearer, also did the folder move outside of goland which avoided the annoying package import changes.  There are three separate commits to also help with seeing code changes verses generated changes.

**Comment by rawlingsj** [comment]
@kauana @jsanin did you have any joy at seeing if the `clients` branch includes what you need?

**Comment by kauana** [comment]
Hey @rawlingsj, we haven't tried the new branch yet, but we did learn today that if you are generating clients for cluster scoped resources (such as ClusterSupplyChain) you should have the tag `+genclient:nonNamespaced` added to your resource type like [this example](https://github.com/knative/networking/blob/3f45c8b29bc15c38fc35874dcb350dc52d57ab06/pkg/apis/networking/v1alpha1/domainclaim_types.go#L24-L25). We were using the listers from this PR and our controller wasn't reconciling property; turns out that's what we were missing.

**Comment by rawlingsj** [comment]
OK thanks @kauana - I've added that extra annotation it for the cluster resources, could you take a look and try it out when you get a chance please?  

I'll rebase the PR and switch to ready for review once confirmed it's working as expected.

**Comment by jsanin** [comment]
hey @rawlingsj I am having issues trying this out because we are using Knative sample controller as based to create our controllers. It turns out that this project (cartographer) depends on `k8s.io/client-go v0.22.3 ` and this version has a new method in this interface `DeploymentInterface`. https://github.com/kubernetes/client-go/blob/v0.22.3/kubernetes/typed/apps/v1/deployment.go
```
ApplyScale(ctx context.Context, deploymentName string, scale *applyconfigurationsautoscalingv1.ScaleApplyConfiguration, opts metav1.ApplyOptions) (*autoscalingv1.Scale, error)
```

And the current version of knative serving depends on `k8s.io/client-go v0.21.4` so I have a compilation error:

```
# knative.dev/pkg/client/injection/kube/client
vendor/knative.dev/pkg/client/injection/kube/client/client.go:1236:9: cannot use &wrapAppsV1DeploymentImpl{...} (type *wrapAppsV1DeploymentImpl) as type "k8s.io/client-go/kubernetes/typed/apps/v1".DeploymentInterface in return argument:
	*wrapAppsV1DeploymentImpl does not implement "k8s.io/client-go/kubernetes/typed/apps/v1".DeploymentInterface (missing ApplyScale method)
```

I am not sure how to solve this ... I am still working on it

**Comment by rawlingsj** [comment]
Yeah - that's the issue with importing informers etc using go dependencies.  You should be able to fix client-go versions using a `replace` in your go.mod file.  Alternatively if that becomes a pain you could try avoiding typed informers and use the dynamic Kubernetes client.

**Comment by jsanin** [comment]
@rawlingsj thanks for the tip about `replace`. So it now compiles. For future reference this is what we add to our project:
```
replace (
	github.com/vmware-tanzu/cartographer => github.com/rawlingsj/cartographer v0.0.7-0.20211117083742-bdefc96eef5c
	k8s.io/api => k8s.io/api v0.21.4
	k8s.io/apimachinery => k8s.io/apimachinery v0.21.4
	k8s.io/client-go => k8s.io/client-go v0.21.4
)
```

we have not been able to test it yet. As the listers, clientset and informers need to be re-generated after adding the `// +genclient:nonNamespaced`

We still see the cluster supply chain lister as namespaced. I can see that in this interface
```
type ClusterSupplyChainLister interface {
	// List lists all ClusterSupplyChains in the indexer.
	// Objects returned here must be treated as read-only.
	List(selector labels.Selector) (ret []*v1alpha1.ClusterSupplyChain, err error)
	// ClusterSupplyChains returns an object that can list and get ClusterSupplyChains.
	ClusterSupplyChains(namespace string) ClusterSupplyChainNamespaceLister
	ClusterSupplyChainListerExpansion
}
```

Will you re-run the code generation and push a new version ? thanks a lot

**Comment by squeedee** [comment]
@rawlingsj Would love to see a `make gen-client` target that will generate new client code when the api sources change.

**Comment by squeedee** [comment]
@rawlingsj A refactor that we've long talked about, to dry up some of the common code in all our controller reconcilers, would greatly benefit from our crd objects implementing client.Object. Seeing as this branch gives us that, I was wondering what the current timeline looks like for getting it out of 'draft' ? What are the blockers and how can I help get us past them?

**Comment by rawlingsj** [comment]
> what the current timeline looks like for getting it out of 'draft' ? What are the blockers and how can I help get us past them?

sounds good - yeah I've pinged @kauana asking for some feedback to confirm the generated clients did what was needed.  My plan was once that's confirmed to rebase, (add the make target from comment above), and take it off draft.

So nothing really left to do other than confirming from @kauana or @jsanin their needs are addressed with it

**Comment by jsanin** [comment]
Hi there. I have been busy getting some things done before the code freeze. I should be able to review this by Monday. Sorry for the delay

**Comment by rawlingsj** [comment]
Rebased pull request and added the `make gen-client` target to simplify generating clients, fakes and listers in the future.

Ideally we should probably still wait for @jsanin or @kauana to confirm their use of the clients is working as expected but I'll take off draft while we wait.

**Comment by squeedee** [comment]
`make gen-client` fails for me with errors like:

```
Generating clientset for carto:v1alpha1 at github.com/vmware-tanzu/cartographer/pkg/generated/clientset
F1130 20:11:41.172228    2391 main.go:64] Error: Failed executing generator: some packages had errors:
errors in package "github.com/vmware-tanzu/cartographer/pkg/generated/clientset/versioned":
open ../../../github.com/vmware-tanzu/cartographer/pkg/generated/clientset/versioned/doc.go: no such file or directory
open ../../../github.com/vmware-tanzu/cartographer/pkg/generated/clientset/versioned/clientset.go: no such file or directory

.....

goroutine 1 [running]:
k8s.io/klog/v2.stacks(0x1)
	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1026 +0x8a
k8s.io/klog/v2.(*loggingT).output(0x15cdc20, 0x3, {0x0, 0x0}, 0xc0135f7dc0, 0x0, {0x1448533, 0xc013c61b30}, 0x0, 0x0)
	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:975 +0x63d
k8s.io/klog/v2.(*loggingT).printf(0x6, 0x1376190, {0x0, 0x0}, {0x0, 0x0}, {0x1340b86, 0x9}, {0xc013c61b30, 0x1, ...})
	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:753 +0x1e5
k8s.io/klog/v2.Fatalf(...)
	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1514
main.main()
	/Users/vmware/go/pkg/mod/k8s.io/code-generator@v0.22.4/cmd/client-gen/main.go:64 +0x38e

goroutine 6 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x0)
	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1169 +0x6a
created by k8s.io/klog/v2.init.0
	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:420 +0xfb
make: *** [gen-client] Error 255
```

**Comment by kauana** [comment]
> `make gen-client` fails for me with errors like:
> 
> ```
> Generating clientset for carto:v1alpha1 at github.com/vmware-tanzu/cartographer/pkg/generated/clientset
> F1130 20:11:41.172228    2391 main.go:64] Error: Failed executing generator: some packages had errors:
> errors in package "github.com/vmware-tanzu/cartographer/pkg/generated/clientset/versioned":
> open ../../../github.com/vmware-tanzu/cartographer/pkg/generated/clientset/versioned/doc.go: no such file or directory
> open ../../../github.com/vmware-tanzu/cartographer/pkg/generated/clientset/versioned/clientset.go: no such file or directory
> 
> .....
> 
> goroutine 1 [running]:
> k8s.io/klog/v2.stacks(0x1)
> 	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1026 +0x8a
> k8s.io/klog/v2.(*loggingT).output(0x15cdc20, 0x3, {0x0, 0x0}, 0xc0135f7dc0, 0x0, {0x1448533, 0xc013c61b30}, 0x0, 0x0)
> 	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:975 +0x63d
> k8s.io/klog/v2.(*loggingT).printf(0x6, 0x1376190, {0x0, 0x0}, {0x0, 0x0}, {0x1340b86, 0x9}, {0xc013c61b30, 0x1, ...})
> 	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:753 +0x1e5
> k8s.io/klog/v2.Fatalf(...)
> 	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1514
> main.main()
> 	/Users/vmware/go/pkg/mod/k8s.io/code-generator@v0.22.4/cmd/client-gen/main.go:64 +0x38e
> 
> goroutine 6 [chan receive]:
> k8s.io/klog/v2.(*loggingT).flushDaemon(0x0)
> 	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1169 +0x6a
> created by k8s.io/klog/v2.init.0
> 	/Users/vmware/go/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:420 +0xfb
> make: *** [gen-client] Error 255
> ```

That's the exact same error I got when I try running it

**Review by emmjohnson — CHANGES_REQUESTED** [review]
I think we want to hold this change for the next release. Putting a blocking review so that we do not accidentally merge it.

**Comment by suarezjulian** [comment]
This is actually something we are now doing ourselves in our projects ( only client-gen though) in our team. It will be great if we can get this merged to use this instead

**Comment by scothis** [comment]
This is an unpopular opinion, but depending on foreign types and clients for k8s resources is ["considered harmful"](https://meyerweb.com/eric/comment/chech.html).

There are frequent (about once a year) upstream api breakages in k8s that force all consumers to be either before the break or after the break. You cannot mix. Depending on those foreign types/clients also means you depend on their transitive dependencies. If there's a bug that's fixed on one side of the break, but you also depend on a component that hasn't updated, tough.

A pattern that works for me is to copy the structs for types you care about (or reverse them from a structural schema) and use the controller-runtime client.

**Comment by squeedee** [comment]
@rawlingsj can we take @scothis last piece of advise, and in doing so close this PR as wont-merge?

**Pull request closed without merging**